configmerge/reference.go

```diff
@@ -0,0 +1,69 @@
+package configmerge
+
+import (
+	"fmt"
+	"path/filepath"
+)
+
+type ConfigReference struct {
+	Repository string `yaml:"repository" json:"repository"`
+	Branch     string `yaml:"branch" json:"branch"`
+	Commit     string `yaml:"commit" json:"commit"`
+	Tag        string `yaml:"tag" json:"tag"`
+	Path       string `yaml:"path" json:"path"`
+}
+
+func (r ConfigReference) Key() string {
+	key := r.Path
+	if r.Repository != "" {
+		key = "repo:" + r.Repository + "," + r.Path
+	}
+
+	if r.Commit != "" {
+		key += "@commit:" + r.Commit
+	} else if r.Tag != "" {
+		key += "@tag:" + r.Tag
+	} else if r.Branch != "" {
+		key += "@branch:" + r.Branch
+	}
+
+	return key
+}
+
+func (r ConfigReference) Validate() error {
+	key := r.Key()
+
+	includePath := r.Path
+	if includePath == "" {
+		return fmt.Errorf("missing YML path in reference: %s", key)
+	}
+
+	if filepath.Ext(includePath) != ".yml" && filepath.Ext(includePath) != ".yaml" {
+		return fmt.Errorf("invalid YML path in reference (%s): %s is not a yaml file", key, includePath)
+	}
+
+	includeCommit := r.Commit
+	isCommitValid := true
+	if includeCommit != "" {
+		isCommitValid = false
+
+		if len(includeCommit) > 5 && len(includeCommit) < 9 {
+			isCommitValid = true
+		} else if len(includeCommit) == 40 {
+			isCommitValid = true
+		}
+	}
+	if !isCommitValid {
+		return fmt.Errorf("invalid commit hash in reference (%s): %s", key, includeCommit)
+	}
+
```

Comment:
Do we need to validate commit hashes or we can leave this job to the entity which tries to clone the repo?

Reply:
This is a usual config model validation, to fail fast the merge process if the reference is in an invalid format.